File: elixir/lib/2024/04_royal_smiths_puzzle/ec202404.ex
defmodule EverybodyCodes2024.Day04 do
  @moduledoc """
  Everybody Codes 2024, day 4: Royal Smith's Puzzle
  """
  require EverybodyCodes

  @doc """
  Parse the input data
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("\n", trim: true) |> Enum.map(&String.to_integer/1)
  end

  @doc """
  Solve part 1
  """
  def part1(puzzle_input) do
    puzzle_input
    |> parse()
    |> then(fn nails -> count_strikes(nails, Enum.min(nails)) end)
  end

  @doc """
  Solve part 2
  """
  def part2(puzzle_input), do: part1(puzzle_input)

  @doc """
  Solve part 3
  """
  def part3(puzzle_input) do
    nails = puzzle_input |> parse()

    {_, approx_target} =
      Enum.min(nails)..Enum.max(nails)//1000
      |> Task.async_stream(fn target -> {count_strikes(nails, target), target} end)
      |> Enum.map(fn {:ok, result} -> result end)
      |> Enum.min()

    (approx_target - 1000)..(approx_target + 1000)
    |> Task.async_stream(&count_strikes(nails, &1))
    |> Enum.map(fn {:ok, result} -> result end)
    |> Enum.min()
  end

  @doc """
  Count the number of hammer strikes needed to align nails at target

  ## Example

      iex> count_strikes([2, 8, 1, 1, 9, 7, 7], 3)
      24
  """
  def count_strikes(nails, target) do
    nails |> Enum.map(&abs(&1 - target)) |> Enum.sum()
  end

  def main(path1, path2, path3) do
    EverybodyCodes.solve(&part1/1, &part2/1, &part3/1, path1, path2, path3)
  end
end

File: python/src/2024_the-kingdom-of-algorithmia/04_royal-smiths-puzzle/ec202404.py
"""Everybody Codes quest 4, 2024: Royal Smith's Puzzle."""


def part1(puzzle_input: str) -> int:
    """Solve part 1."""
    nails = [int(nail) for nail in puzzle_input.split()]
    return sum(nails) - min(nails) * len(nails)


def part2(puzzle_input: str) -> int:
    """Solve part 2."""
    return part1(puzzle_input)


def part3(puzzle_input: str) -> int:
    """Solve part 3."""
    nails = [int(nail) for nail in puzzle_input.split()]

    # Search for best target, start at approximate mean
    target = sum(nails) // len(nails)
    min_strikes = count_strikes(nails, target)
    step = -(2**12) if count_strikes(nails, target - 1) < min_strikes else 2**12
    while True:
        target += step
        num_strikes = count_strikes(nails, target)
        if num_strikes > min_strikes:
            if abs(step) == 1:
                return min_strikes
            else:
                step //= -8
        min_strikes = num_strikes


def count_strikes(nails: list[int], target: int) -> int:
    """Count the number of strikes necessary for all nails to reach target length.

    ## Example

    >>> count_strikes([3, 4, 7, 8], target=3)
    10
    >>> count_strikes([2, 4, 5, 6, 8], target=5)
    8
    >>> count_strikes([12, 4, 6, 9, 7, 5, 5], target=4)
    20
    >>> count_strikes([12, 4, 6, 9, 7, 5, 5], target=6)
    14
    """
    return sum(abs(nail - target) for nail in nails)
